### PR TITLE
V.1.1.0 release

### PR DESCRIPTION
**Pull request by JoshC8C7**
Support for inference using Snowpark Container Services.
Support for using chat templates as input
Support for generating output adhering to a specified Regular Expression or JSON Schema.
Support for versions >=12.6.x

### CODE REVIEW

File: docs/CHANGELOG.MD
Comment (adamnellis) [inline docs/CHANGELOG.MD]: [minor] Make it clear that this version number refers to DSS (as opposed to, say, Takeoff, or Snowflake)

```suggestion
- Support for DSS versions >=12.6.x
```

File: java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java
Comment (adamnellis) [inline java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java]: [minor] `tokenClient` appears to be only used within this `if` statement, so it could be a local variable rather than an instance variable.

File: java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java
Comment (adamnellis) [inline java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java]: [suggestion] 

This log message is confusing, because the code doesn't know whether we should have the snowflake params set or not.

As far as I can tell, Snowflake is one thing that this plugin can support, but is not the main use case. 

We could separate out the Snowflake use case from the non-snowflake use case by using two different model types. This would clean the code and reduce some possibilities for errors.

---

To do this, you would need to:
- Create a second directory under `java-llms`, containing a second class extending `CustomLLMClient`, and a second `llm.json` file
- In the `llm.json` files, have one of them define the Snowflake params and one not
    - And name them differently, by changing the `meta label`
    - Remember to update the `id` and `clientClass` properties in `llm.json` too
- In the Java files, you can re-use most of the code by having them extend a common base class, or maybe better, having the Snowflake version extend the non-Snowflake one
    - I think it's just the `init` method that the Snowflake case needs to override, calling `super` and then adding the Snowflake-specific authorization header to `this.client` 

If you have any questions about this, feel free to ask me 🙂 

---

The benefits of this would be:
- The Snowflake-specific code would be all in one place
- The non-Snowflake case would be not cluttered with Snowflake code
- In the Dataiku UI, the user would see two options under the `type` dropdown when adding a model. This is where they would choose whether they want to use Snowflake or not
- For users who are not using Snowflake, they would not see the `Snowflake preset` and `Snowflake account url` params in the UI
- The help messages in the UI could be made specific to each case 
    - e.g. currently, the `Inference endpoint URL` help text describes both cases, but it would only need to describe the one case that the user has selected

---

I think this would be a useful improvement for not much effort, but I understand if it's not something you want to do. 

841C9924-2B4C-4D23-B11C-F4D56A711251_1_201_a

**Review by adamnellis — COMMENTED** [review]
- ✅ I've built the plugin and imported it into DSS - all working there
- I haven't run any models yet
- Below are some initial observations from reading the code

File: java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java
Comment (adamnellis) [inline java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java]: This doesn't appear to work as you expect.

When I don't enter anything for consumer group, I get a NPE here, because `settings.config.get("consumer_group")` is `null`.

Do you get the same result? 
- I'm hoping isn't not that different versions of the `gson` library behave differently.

**Review by adamnellis — CHANGES_REQUESTED** [review]
✅ Tested the non-snowflake version on an AWS-deployed version of Takeoff that I set up myself.

✅ Tested a chat completion model 

⚠️ Found a small NPE bug, detailed below

---

I also had a look at your docs at https://docs.titanml.co/docs/Docs/integrations/dataiku/

I think the example you give in the "Chat Templates" section might be a little too simple, because this case can be achieved in DSS natively, without needing chat templates in this plugin. The screenshot you show has text boxes for system prompt and user prompt. 

Screenshot 2024-08-16 at 15 53 50

There are definitely cases that this chat template can implement that DSS visual recipes cannot, so I think it would make your product look better if you included a more complex example. 

For example, DSS visual recipes cannot include input fields in messages that are not the last one. For example, you could do something like this:
```json
[
    {"role": "system", "content": "You are a helpful assistant whose job is to explain sayings. Make sure you answer truthfully, and do not include any {{things_to_not_include}} in your response."},
    {"role": "user", "content": "What does this saying mean? {{example_saying_1}}"}
    {"role": "assistant", "content": "I can help with that, it means: {{saying_explanation_1}}"}
    {"role": "user", "content": "Explain what this means: {{example_saying_2}}"}
    {"role": "assistant", "content": "I can help with that, it means: {{saying_explanation_2}}"}
    {"role": "user", "content": "{{saying_to_explain}}"}
]
```

**Review by adamnellis — CHANGES_REQUESTED** [review]
✅ Tested the snowflake code path, using Titan's Snowflake instance, all working well.

🔴 I think it's worth fixing the NPE bug mentioned above, and the small things, as they are easy to fix.

🟡 I leave it up to you whether you want to do the slightly bigger change of moving the Snowflake parts into a separate model type.

**Comment by JoshC8C7** [comment]
Hi @adamnellis thanks for the review - I've addressed your suggestions, and will update our own docs as you suggested. Thanks!

File: java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java
Comment (adamnellis) [inline java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java]: [nitpick] 

I like the default `consumer_group`, this is a nice addition 👍 

I always like to make it clear in the logs when we modify the user's input.

```suggestion
            logger.info("No consumer group was specified, defaulting to 'primary'");
            consumer_group = "primary";
```

File: java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java
Comment (adamnellis) [inline java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java]:
```suggestion
            logger.info("No consumer group was specified, defaulting to 'primary'");
            consumer_group = "primary";
```

File: java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java
Comment (adamnellis) [inline java-llms/titanMLConnector/com/titanml/llm/TitanMLLLMConnector.java]:
```suggestion
            logger.info("No consumer group was specified, defaulting to 'primary'");
            consumer_group = "primary";
```

**Review by adamnellis — APPROVED** [review]
✅ Thank your for all the changes - they look good to me.

✅ I like your addition of making a blank consumer group default to `primary`.

✅ I re-tested it, and all still works.

🟡 I spotted a tiny improvement that could be made to a log message. I leave it up to you whether you want to change that. (Doesn't need a re-review from me if you change it!)

🟡 Note that this PR is set to merge into the `v1.0.0-release` branch. I presume you don't actually want to do this merge, and this PR is just for me to see the differences between the current version and the previous version.

**Comment by JoshC8C7** [comment]
Yeah the pr is just to indicate the changes - I was a bit overzealous in previously merging to main haha

**Pull request merged**
